proc_macros/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+            }
+
+            pub fn values(&self) -> #values_struct_name<'_, T> {
+                #values_struct_name::new(self)
+            }
+        }
+
+        pub struct #iter_struct_name<'collection, T> {
+            collection: &'collection #collection_type_name<T>,
+            next_index: usize,
+        }
+
+        impl<'collection, T> #iter_struct_name<'collection, T> {
+            pub fn new(collection: &'collection #collection_type_name<T>) -> Self {
+                Self {
+                    collection,
+                    next_index: 0,
+                }
+            }
+        }
+
+        impl<'collection, T> Iterator for #iter_struct_name<'collection, T> {
+            type Item = (#name, &'collection T);
+
+            fn next(&mut self) -> Option<Self::Item> {
+                if self.next_index < self.collection.len() {
+                    let ret = Some((
+                        #all_variants_collection_name[self.next_index],
+                        &self.collection.0[self.next_index],
+                    ));
+                    self.next_index += 1;
+                    ret
+                } else {
+                    None
+                }
+            }
+        }
+
+        pub struct #values_struct_name<'collection, T> {
+            collection: &'collection #collection_type_name<T>,
+            next_index: usize,
+        }
+
+        impl<'collection, T> #values_struct_name<'collection, T> {
+            pub fn new(collection: &'collection #collection_type_name<T>) -> Self {
+                Self {
+                    collection,
+                    next_index: 0,
+                }
+            }
+        }
+
+        impl<'collection, T> Iterator for #values_struct_name<'collection, T> {
+            type Item = &'collection T;
+
+            fn next(&mut self) -> Option<Self::Item> {
+                if self.next_index < self.collection.len() {
+                    let ret = Some(&self.collection.0[self.next_index]);
+                    self.next_index += 1;
+                    ret
+                } else {
+                    None
+                }
+            }
+        }
+    }
+}
+
+fn get_index_implementations(
+    name: &Ident,
+    collection_type_name: &Ident,
+) -> proc_macro2::TokenStream {
+    quote! {
+        impl<T> Index<#name> for #collection_type_name<T> {
+            type Output = T;
+
+            fn index(&self, index: #name) -> &Self::Output {
+                &self.0[index as usize]
+            }
+        }
+
+        impl<T> Index<usize> for #collection_type_name<T> {
+            type Output = T;
+
+            fn index(&self, index: usize) -> &Self::Output {
+                &self.0[index]
+            }
+        }
+    }
+}
+
+fn get_deref_implementation(
+    collection_type_name: &Ident,
+    variants: &[Ident],
+) -> proc_macro2::TokenStream {
+    let len = variants.len();
+    quote! {
+        impl<T> Deref for #collection_type_name<T> {
```

Comment:
Added a `Deref` implementation along with the newtype wrapper (I think it's currently only getting used to call the underlying `.len()`)